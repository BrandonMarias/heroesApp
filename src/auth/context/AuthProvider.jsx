import React, { useReducer } from "react";
import { AuthContext } from "../context/AuthContext";
import { authReducer } from "./authReducer";
import { types } from "../types/types";

const init = () => {
    const user = JSON.parse(localStorage.getItem("user"));

    return {
        logged: !!user,
        ...user,
    };
};

export const AuthProvider = ({ children }) => {
    const [user, dispatch] = useReducer(authReducer, {}, init);

    const login = (name = "") => {
        const user = { id: "ABC", name };
        const action = { type: types.login, payload: user };
        localStorage.setItem("user", JSON.stringify(user));
        dispatch(action);
    };

    const logout = () => {
        localStorage.removeItem("user");
        const action = { type: types.logout}
        console.log(action)
        dispatch(action)
    }

    return (
        <AuthContext.Provider value={{ user, login, logout }}>
            {children}
        </AuthContext.Provider>
    );
};
