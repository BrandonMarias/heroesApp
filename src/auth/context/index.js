export * from "./AuthContext";
export * from "./types"
export * from "./authReducer"