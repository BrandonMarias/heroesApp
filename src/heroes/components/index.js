export * from "./NavBar";
export * from "./ButtonNavToggler";
export * from "./HeroCard";
export * from "./HeroesList";
export * from "./SearchForm";
