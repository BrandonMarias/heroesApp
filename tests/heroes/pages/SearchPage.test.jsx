import { fireEvent, render, screen } from "@testing-library/react";
import { SearchPage } from "../../../src/heroes";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "../../../src/auth/context/AuthContext";
import { AppRouter } from "../../../src/router/AppRouter";

const MockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => MockedUseNavigate,
}));

describe("Pruebas en <SearchPage", () => {

    beforeEach(() => {
        MockedUseNavigate.mockClear();
    });

    test("snapshot", () => {
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();

        screen.debug();
    });

    test("should batman and the query with the input string", () => {
        render(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <SearchPage />
            </MemoryRouter>
        );

        const image = screen.getByRole("img");
        expect(image.src).toContain("batman");
        expect(screen.getByText("Batman")).toBeTruthy();
    });

    test("should show an error if there is no hero", () => {
        render(
            <MemoryRouter initialEntries={["/search?q=batman123"]}>
                <SearchPage />
            </MemoryRouter>
        );

        expect(
            screen.getByText("There is no a hero with batman123")
        ).toBeTruthy();
    });

    test("should call the getHeroByQuery with the correct argument", () => {
        const contextValue = {
            user: {
                logged: true,
                name: "Fernando",
            },
        };

        render(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter initialEntries={["/search?q=batman"]}>
                    <AppRouter />
                </MemoryRouter>
            </AuthContext.Provider>
        );

        const input = screen.getByRole("searchbox");

        fireEvent.change(input, { target: { value: "superman" } });

        expect(input.value).toBe("superman");

        const form = screen.getByRole("form");

        fireEvent.submit(form);

        expect(MockedUseNavigate).toHaveBeenCalledWith( "./search?q=superman" );

        screen.debug();
    });
});
