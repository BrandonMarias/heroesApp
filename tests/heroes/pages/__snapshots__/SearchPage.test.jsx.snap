// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage snapshot 1`] = `
<div>
  <h1
    class="mt-4"
  >
    Results
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-md-7"
    >
      <div
        class="alert alert-info animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
    </div>
  </div>
  <br />
</div>
`;
